### ISSUE TITLE
[BUG] extreme memory leak in Maintenance process.

### ISSUE THREAD

**Issue by whytewolf** [saltstack/salt#58785]
**Description**
salt 3002 just installed last night. woke up this morning to find all memory and most of the swap taken up. checked `ps aux | grep salt` before restarting service to get memory back. found that the maintenance processes was the only large memory hog. 

**Setup**
install 3002 and wait.

**Steps to Reproduce the behavior**
install 3002 and have minions. 

**Expected behavior**
I have memory on my system after only a day of having a newly updated system.

**Screenshots**
![mem leak](https://user-images.githubusercontent.com/1143644/96896123-ffaeaf00-1441-11eb-97cd-d30c0bab8eb6.jpg)

the little uptick at 18:00 is when I installed 3002. 

**Versions Report**
<details><summary>salt --versions-report</summary>
(Provided by running salt --versions-report. Please also mention any differences in master/minion versions.) 

```
Salt Version:                                                                    
           Salt: 3002                                                            
                                                                                 
Dependency Versions:                                                             
           cffi: 1.14.0                                                          
       cherrypy: unknown                                                         
       dateutil: Not Installed                                                   
      docker-py: Not Installed                                                   
          gitdb: Not Installed                                                   
      gitpython: Not Installed                                                   
         Jinja2: 2.11.1                                                          
        libgit2: 0.26.6                                                          
       M2Crypto: 0.35.2                                                          
           Mako: Not Installed                                                   
   msgpack-pure: Not Installed                                                   
 msgpack-python: 1.0.0                                                           
   mysql-python: Not Installed                                                   
      pycparser: 2.20                                                            
       pycrypto: Not Installed                                                   
   pycryptodome: 3.9.8                                                           
         pygit2: 0.26.4                                                          
         Python: 3.6.8 (default, Apr  2 2020, 13:34:55)                          
   python-gnupg: 2.3.1                                                           
         PyYAML: 3.13                                                            
          PyZMQ: 17.0.0                                                          
          smmap: Not Installed                                                   
        timelib: Not Installed                                                   
        Tornado: 4.5.3                                                           
            ZMQ: 4.1.4                                                           
                                                                                 
System Versions:                                                                 
           dist: centos 7 Core                                                   
         locale: UTF-8                                                           
        machine: x86_64                                                          
        release: 3.10.0-1127.13.1.el7.x86_64                                     
         system: Linux                                                           
        version: CentOS Linux 7 Core                                             
```
</details>

**Additional context**

**Comment by whytewolf** [comment]
after further testing, this might be related to SSE products' interaction with the maintenance process. The bug could still be in both products so will keep this open for now.

**Comment by whytewolf** [comment]
![sse_mem_leak](https://user-images.githubusercontent.com/1143644/96904992-b6b02800-144c-11eb-8de3-0b281f389634.jpg)

screenshot from another master that had sse enabled then I disabled it.

**Comment by whytewolf** [comment]
so after leaving one of my masters out of SSE for a little bit it looks like most of the leak is with SSE. however, I am still seeing a small bit of leak almost unnoticeable without sse running. want to give it more time maybe 24 hours to be sure.

**Comment by tuxthepenguin84** [comment]
I can confirm I'm seeing similar behavior as well on my Salt Master (3002), just upgraded today and ran into the same issue.

**Comment by whytewolf** [comment]
ok, After leaving my system up for 24 hours without SSE. There is still a memory leak. on this system, it is slow but present. since this system I'm testing with only has about 7 minions that could explain why it is slow. With more minions, I can see this being a much faster leak. This system also doesn't have git_pillar like an earlier system I saw this on. it has no extra bells or whistles setup.

**Comment by Arderos** [comment]
I can confirm this behavior with salt-master 3002 and Ubuntu 18.04
![image](https://user-images.githubusercontent.com/12333169/97017515-b0a66e00-1556-11eb-91c7-a3159225b40f.png)

**Comment by jbouter** [comment]
Can confirm with `salt-master` 3002 and Ubuntu 20.04 LTS

![Screenshot from 2020-10-24 10-47-21](https://user-images.githubusercontent.com/6340770/97077738-6d93db80-15e6-11eb-9d57-a30ebb847b96.png)

**Comment by jeffdyke** [comment]
There are enough fancy graphs (thanks folks!), i have 3 salt masters all 3002, on ubuntu 20.04 EC2 t3.small instances, which has always been sufficient.  Prod has 20 minions, Staging  has 9, and dev varies but right now is down to 4.  All three had OOM errors and the OS reaped the process on each server.  All salt master instances, run salt master/minion, cloudwatch agent, wazuh-agent and nothing else.  I rebooted the dev version (4 minions) last night and right now its up to 60% of the memory is taken by salt master. I've set up a cron for these systems to restart the master at 90%, as it completely took down the instance in one environment and i could only be restarted via aws console/cli.

Just adding this b/c i don't think it has to do with the # of minions, or other software interfering.

**Comment by remyd1** [comment]
Hi,
I can confirm this bug, on processors too. I upgraded the salt-master yesterday; this morning, memory was full and the 4 processors were loaded at 100%.

I have 77 minions and many crons tasks.

versions-report
```
Salt Version:
           Salt: 3002
 
Dependency Versions:
           cffi: 1.14.3
       cherrypy: Not Installed
       dateutil: 2.7.3
      docker-py: Not Installed
          gitdb: 2.0.6
      gitpython: 3.0.7
         Jinja2: 2.10.1
        libgit2: 1.0.1
       M2Crypto: Not Installed
           Mako: Not Installed
   msgpack-pure: Not Installed
 msgpack-python: 0.6.2
   mysql-python: Not Installed
      pycparser: 2.20
       pycrypto: 2.6.1
   pycryptodome: 3.6.1
         pygit2: 1.3.0
         Python: 3.8.5 (default, Jul 28 2020, 12:59:40)
   python-gnupg: 0.4.5
         PyYAML: 5.3.1
          PyZMQ: 18.1.1
          smmap: 2.0.5
        timelib: Not Installed
        Tornado: 4.5.3
            ZMQ: 4.3.2
 
System Versions:
           dist: ubuntu 20.04 focal
         locale: utf-8
        machine: x86_64
        release: 5.4.0-52-generic
         system: Linux
        version: Ubuntu 20.04 focal
```

**Comment by xcorvis** [comment]
Also seeing this on 3002.1 on Ubuntu 16.04. I have about 1700 minions and my salt master blows through 24 GB of RAM in about 15 hours and starts doing oomkills. In 3001 it was steady at ~19 GB. Not seeing CPUs max out though, they stay relatively low.

<details>
  <summary>salt -V</summary>

```
Salt Version:
           Salt: 3002.1
 
Dependency Versions:
           cffi: Not Installed
       cherrypy: Not Installed
       dateutil: 2.4.2
      docker-py: Not Installed
          gitdb: 0.6.4
      gitpython: 1.0.1
         Jinja2: 2.8
        libgit2: Not Installed
       M2Crypto: Not Installed
           Mako: Not Installed
   msgpack-pure: Not Installed
 msgpack-python: 0.6.2
   mysql-python: Not Installed
      pycparser: Not Installed
       pycrypto: 2.6.1
   pycryptodome: 3.4.7
         pygit2: Not Installed
         Python: 3.5.2 (default, Oct  7 2020, 17:19:02)
   python-gnupg: 0.3.8
         PyYAML: 3.11
          PyZMQ: 17.1.2
          smmap: 0.9.0
        timelib: Not Installed
        Tornado: 4.5.3
            ZMQ: 4.1.4
 
System Versions:
           dist: ubuntu 16.04 Xenial Xerus
         locale: UTF-8
        machine: x86_64
        release: 4.4.0-189-generic
         system: Linux
        version: Ubuntu 16.04 Xenial Xerus
```

</details>

**Comment by litnimax** [comment]
We have memory leak for salt-minion KeepAlive MultiMinionProcessManager MinionProcessManager.
Does it relate to this issue?

**Comment by mirceaulinic** [comment]
Did anyone notice any improvement at all in 3002.2 after #58862? Thanks in
advance for information!

**Comment by jeffdyke** [comment]
For @litnimax and @mirceaulinic .  I have not seen this problem after an upgrade to the official release of 3002.02.  Unfortunately i forget if i purged anything or reinstalled, but i don't think so, as i would normally report that back.  Of course it could still happen, this is my setup right now, and all is fine:

```
alt Version:
          Salt: 3002.2
 
Dependency Versions:
          cffi: Not Installed
      cherrypy: Not Installed
      dateutil: 2.7.3
     docker-py: Not Installed
         gitdb: 2.0.6
     gitpython: 3.0.7
        Jinja2: 2.10.1
       libgit2: Not Installed
      M2Crypto: Not Installed
          Mako: Not Installed
       msgpack: 0.6.2
  msgpack-pure: Not Installed
  mysql-python: Not Installed
     pycparser: Not Installed
      pycrypto: Not Installed
  pycryptodome: 3.6.1
        pygit2: Not Installed
        Python: 3.8.5 (default, Jul 28 2020, 12:59:40)
  python-gnupg: 0.4.5
        PyYAML: 5.3.1
         PyZMQ: 18.1.1
         smmap: 2.0.5
       timelib: Not Installed
       Tornado: 4.5.3
           ZMQ: 4.3.2
 
System Versions:
          dist: ubuntu 20.04 focal
        locale: utf-8
       machine: x86_64
       release: 5.4.0-1037-aws
        system: Linux
       version: Ubuntu 20.04 focal

```

**Comment by sagetherage** [comment]
closing as addressed in v3002.2 - with more work to be done to continue to improve
